### PR TITLE
[IMP] project: Implement favorite functionality in project.

### PR DESCRIPTION
**Pull request by dpa-odoo**
Task : https://www.odoo.com/web#id=21497&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720

Pad : https://pad.odoo.com/p/r.Z3TCkqlxFRuxtGpq

### CODE REVIEW

File: addons/project/project.py [hunk 1]
```diff
@@ -122,6 +122,40 @@ def _get_visibility_selection(self, cr, uid, context=None):
                 ('employees', _('All Employees Project: all employees can access')),
                 ('followers', _('Private Project: followers only'))]
 
+    def _get_favorite(self, cr, uid, ids, name, args, context=None):
+        res = {}
+        for member in self.browse(cr, uid, ids, context=context):
```
Comment (atp-odoo) [inline addons/project/project.py]: give proper name 'project' because it's project instance not member.
for project in self.browse(...

Comment (dpa-odoo) [inline addons/project/project.py]: Ok...Done.

File: addons/project/project.py [hunk 2]
```diff
@@ -122,6 +122,24 @@ def _get_visibility_selection(self, cr, uid, context=None):
...
+    def _get_favorite(self, cr, uid, ids, name, args, context=None):
+        res = {}
+        for project in self.browse(cr, uid, ids, context=context):
+            res[project.id] = False
+            if uid in project.member_ids.ids:
+                res[project.id] = True
+        return res
+
+    def _set_favorite(self, cr, uid, ids, field_name, field_value, args, context=None):
+        members = self.browse(cr, uid, ids, context=context).member_ids.ids
```
Comment (atp-odoo) [inline addons/project/project.py]: change variable name : member_ids and iteration required.

File: addons/project/project.py [hunk 3]
```diff
@@ -203,8 +223,13 @@ def attachment_tree_view(self, cr, uid, ids, context):
         'user_id': lambda self,cr,uid,ctx: uid,
         'alias_model': 'project.task',
         'privacy_visibility': 'employees',
+        'member_ids': _default_user_ids,
     }
 
+    def toggle_favorite(self, cr, uid, ids, context=None):
+        project = self.browse(cr, uid, ids, context=context)
```
Comment (atp-odoo) [inline addons/project/project.py]: variable name should be projects

File: addons/project/project.py [hunk 4]
```diff
@@ -122,6 +122,24 @@ def _get_visibility_selection(self, cr, uid, context=None):
...
+            res[project.id] = False
+            if uid in project.member_ids.ids:
+                res[project.id] = True
+        return res
+
+    def _set_favorite(self, cr, uid, ids, field_name, field_value, args, context=None):
+        if field_value:
+            for project in self.browse(cr, uid, ids, context=context):
+                member_ids = project.member_ids.ids
+                project.write({'member_ids': [(6, 0, member_ids + [uid])]})
```
Comment (tde-banana-odoo) [inline addons/project/project.py]: There is a command for that. I let you find it.

File: addons/project/project.py [hunk 5]
```diff
@@ -191,6 +209,8 @@ def attachment_tree_view(self, cr, uid, ids, context):
         ),
         'date_start': fields.date('Start Date'),
         'date': fields.date('Expiration Date', select=True, track_visibility='onchange'),
+        'show_on_dashboard': fields.function(_get_favorite, fnct_inv=_set_favorite, type="boolean", string='Show Projetc on dashboard', help="Whether this project should be displayed on the dashboard or not"),
```
Comment (tde-banana-odoo) [inline addons/project/project.py]: Name should be more consistent. It should be something like is_favorite.

File: addons/project/project.py [hunk 6]
```diff
@@ -122,6 +122,24 @@ def _get_visibility_selection(self, cr, uid, context=None):
                 ('employees', _('All Employees Project: all employees can access')),
                 ('followers', _('Private Project: followers only'))]
 
+    def _get_favorite(self, cr, uid, ids, name, args, context=None):
+        res = {}
+        for project in self.browse(cr, uid, ids, context=context):
+            res[project.id] = False
+            if uid in project.member_ids.ids:
```
Comment (tde-banana-odoo) [inline addons/project/project.py]: You could simplify those lines with a one-liner.

File: addons/project/project.py [hunk 7]
```diff
@@ -203,8 +223,17 @@ def attachment_tree_view(self, cr, uid, ids, context):
...
         'alias_model': 'project.task',
         'privacy_visibility': 'employees',
+        'member_ids': _default_user_ids,
     }
 
+    def toggle_favorite(self, cr, uid, ids, context=None):
+        projects = self.browse(cr, uid, ids, context=context)
+        if uid in projects.member_ids.ids:
+            projects.write({'show_on_dashboard': False})
+            projects.write({'member_ids': [(3, uid)]})
```
Comment (tde-banana-odoo) [inline addons/project/project.py]: Don't you think there is something not necessary here ?

**Pull request merged**